core/sail/memory/src/main/java/org/eclipse/rdf4j/sail/memory/model/MemValueFactory.java

```diff
@@ -228,13 +259,10 @@ public MemResource getOrCreateMemResource(Resource resource) {
 	/**
 	 * See {@link #getOrCreateMemValue(Value)} for description.
 	 */
-	public synchronized MemIRI getOrCreateMemURI(IRI uri) {
-		MemIRI memURI = getMemURI(uri);
-
-		if (memURI == null) {
-			// Namespace strings are relatively large objects and are shared
-			// between uris
+	public MemIRI getOrCreateMemURI(IRI uri) {
+		return iriRegistry.getOrAdd(uri, () -> {
 			String namespace = uri.getNamespace();
+			assert namespace != null;
 			String sharedNamespace = namespaceRegistry.get(namespace);
```

Comment:
Should this also be handled using a `getOrAdd` instead of the `if-else` construction currently used?